marshal/marshal.go

```diff
@@ -0,0 +1,213 @@
+package marshal
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"golang.org/x/exp/maps"
+)
+
+const forceSendFieldName = "ForceSendFields"
+
+// Marshal returns a JSON encoding of the given object. Included fields:
+// - non-empty value
+// - a basic type whose field's name is present in forceSendFields
+// Embedded structs are still considered a separate struct. ForceSendFields
+// in an embedded struct only impact the fields of the embedded struct.
+// Conversely, an embedded struct is not impacted by the ForceSendFields
+// of the struct containing it.
+func Marshal(object any) ([]byte, error) {
+	dataMap, err := structAsMap(object)
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(dataMap)
+}
+
+// Converts the object to a map
+func structAsMap(object any) (map[string]any, error) {
+	result := make(map[string]any)
+
+	// If the object is nil or a pointer to nil, don't do anything
+	if object == nil || (reflect.ValueOf(object).Kind() == reflect.Ptr && reflect.ValueOf(object).IsNil()) {
+		return result, nil
+	}
+
+	value := reflect.ValueOf(object)
+	value = reflect.Indirect(value)
+	objectType := value.Type()
+
+	includeFields, err := getForceSendFields(object, getTypeName(objectType))
+
+	if err != nil {
+		return nil, err
+	}
+
+	fields := *getTypeFields(objectType)
+
+	for i := 0; i < value.NumField(); i++ {
+		jsonTag := fields[i].JsonTag
+		tag, err := parseJSONTag(jsonTag)
+		if err != nil {
+			return nil, err
+		}
+
+		fieldValue := value.Field(i)
+		fieldStruct := fields[i]
+
+		// Anonymous fields should be marshalled using the same JSON, and then merged into the same map
+		if fieldStruct.Anonymous && fieldValue.IsValid() {
+			anonymousFieldResult, err := structAsMap(fieldValue.Interface())
+			if err != nil {
+				return nil, err
+			}
+			result = mergeMaps(result, anonymousFieldResult)
```

Comment:
This now means the anonymous struct takes precedence over the parent, which I think is the reverse of what should actually happen. Can you add a test for this? I suspect we won't hit this path in the SDK because we don't do composition and we don't have field name aliasing.

Reply:
I have the order. In any case, there should not be duplicates. json.marshal doesn't support duplicates. 
If we want to do something, we may want to throw an error instead.